scipy/integrate/_ivp/tests/test_ivp.py

```diff
@@ -758,7 +776,10 @@ def early_event(t, y):
         assert res.t_events[0][0] == 7
 
 
-def test_event_dense_output_LSODA():
+def test_event_dense_output_LSODA(num_parallel_threads):
+    if num_parallel_threads > 1:
+        pytest.skip('LSODA does not allow for concurrent execution')
```

Comment:
Can you make a separate issue for this (assuming we want to fix this)? Is the difference with other solvers that there wasn't a good place to add a lock?

Reply:
It seems that  `lsoda`, `zvode` and `vode` (defined as FORTRAN routines) are not thread-safe, I don't know much about the wrapper generation and how to add locks to their calls, so I deferred the locking to the testing instead.

Reply:
I had a closer look. There is a Python interface that calls the Fortran routines (e.g., `_lsoda.lsoda`). When I remove a test skip I get a crash, that is fixed by putting a lock around that call - here is an example diff:

```diff
diff --git a/scipy/integrate/_ode.py b/scipy/integrate/_ode.py
index 2d52e093e..7dec2dbf5 100644
--- a/scipy/integrate/_ode.py
+++ b/scipy/integrate/_ode.py
@@ -82,6 +82,7 @@ __all__ = ['ode', 'complex_ode']
 
 import re
 import warnings
+from threading import Lock
 
 from numpy import asarray, array, zeros, isscalar, real, imag, vstack
 
@@ -95,6 +96,9 @@ _vode_int_dtype = _vode.types.intvar.dtype
 _lsoda_int_dtype = _lsoda.types.intvar.dtype
 
 
+ODE_LOCK = Lock()
+
+
 # ------------------------------------------------------------------------------
 # User interface
 # ------------------------------------------------------------------------------
@@ -1343,7 +1347,8 @@ class lsoda(IntegratorBase):
             self.acquire_new_handle()
         args = [f, y0, t0, t1] + self.call_args[:-1] + \
                [jac, self.call_args[-1], f_params, 0, jac_params]
-        y1, t, istate = self.runner(*args)
+        with ODE_LOCK:
+            y1, t, istate = self.runner(*args)
         self.istate = istate
         if istate < 0:
             unexpected_istate_msg = f'Unexpected istate={istate:d}'
diff --git a/scipy/integrate/tests/test_integrate.py b/scipy/integrate/tests/test_integrate.py
index f83876d49..0f45b6905 100644
--- a/scipy/integrate/tests/test_integrate.py
+++ b/scipy/integrate/tests/test_integrate.py
@@ -112,11 +112,8 @@ class TestOde(TestODEClass):
                 self._do_problem(problem, 'zvode', 'adams')
             self._do_problem(problem, 'zvode', 'bdf')
 
-    @pytest.mark.thread_unsafe
-    def test_lsoda(self, num_parallel_threads):
+    def test_lsoda(self):
         # Check the lsoda solver
-        if num_parallel_threads > 1:
-            pytest.skip(reason='LSODA does not allow for concurrent execution')
         for problem_cls in PROBLEMS:
             problem = problem_cls()
             if problem.cmplx:
```

The `ode` and `odeint` classes are annoying to read, and the existing `check_handle`/`acquire_new_handle` seems wrong, but a few well-places locks should fix this. Just skipping the tests isn't useful, so let's try to remove those skips again and fix the actual issues.

Reply:
@andfoy this wasn't addressed - I pushed a commit to fix this up. Everything except `integrate.solve_ivp` is now thread-safe.